### PR TITLE
A unit test for TPRs and apply 

### PR DESCRIPTION
**Pull request by ahakanbaba**
**What this PR does / why we need it**:
This PR is the first example for unit tests with apply and ThirdPartyResoruces. 

We have seen that there has been degradations in TPR usage with apply from 1.4.7 onwards to 1.4.8 and 1.5. Hopefully with more test coverage we want to ensure more TPR stability. 

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:
There has been discussions about improving TPR stability with @pwittrock @ymqytw and @huggsboson. This PR is my first attempt to add more dedicated tests following those discussions. 

I have based this test on 1.4.7, because the TPR's are known not to work in later than 1.4.7 releases. I was unsure whether one could implement a passing TPR unit test in later than 1.4.7.  

I see that some of the changes in this PR are not directly mergeable to HEAD. There has been factorizations to the fake library. I am sending out this diff to start the discussion and get some feedback about tests. I will do the modifications later on so this becomes mergeable. 

**Release note**:

```release-note
```

### CODE REVIEW

**Comment by pwittrock** [comment]
Are these already present in master?

**Comment by pwittrock** [comment]
Try this against master first.  I think things are in a much better state there.

**Comment by ahakanbaba** [comment]
These are not present in master and thanks for the suggestion I will try these against master. 

In the meantime I can get some feedback about the approaches I have taken.

**Comment by pwittrock** [comment]
@ymqytw Would you take a review pass on this?

**Comment by ymqytw** [comment]
@pwittrock  Sure.

**Comment by ahakanbaba** [comment]
@ymqytw 
Does this look reasonable ? Should I spend time to port this to master or do some changes first ?

File: pkg/kubectl/cmd/apply_test.go [hunk 1]
```diff
@@ -216,6 +238,62 @@ func TestApplyObject(t *testing.T) {
...
+		Client: fake.CreateHTTPClient(func(req *http.Request) (*http.Response, error) {
+			switch p, m := req.URL.Path, req.Method; {
+			case p == pathTPR && m == "GET":
+				bodyTPR := ioutil.NopCloser(bytes.NewReader(currentTPR))
+				return &http.Response{
+					StatusCode: 200,
+					Header:     defaultHeader(),
+					Body:       bodyTPR,
+				}, nil
+			case p == pathTPR && m == "PATCH":
```
Comment (ymqytw) [inline pkg/kubectl/cmd/apply_test.go]: You are not checking if the patch has the expected field here.
Please take a look this test `func TestApplyNULLPreservation(t *testing.T)`

File: pkg/kubectl/cmd/apply_test.go [hunk 2]
```diff
@@ -216,6 +238,62 @@ func TestApplyObject(t *testing.T) {
...
+
+	buf := bytes.NewBuffer([]byte{})
+
+	cmd := NewCmdApply(f, buf)
+	cmd.Flags().Set("filename", filenameTPR)
+	cmd.Flags().Set("output", "name")
+	cmd.Run(cmd, []string{})
+
+	// Check the output of the kubectl apply command.
+	expectTPR := "thirdpartyresource/" + nameTPR + "\n"
```
Comment (ymqytw) [inline pkg/kubectl/cmd/apply_test.go]: In your case, it should be `"test-tpr/" + nameTPR`, because you are applying on a TPR obj, not the TPR definition.

**Review by mengqiy — COMMENTED** [review]
The idea of adding a unit test for TPR is reasonable. But have 2 comments to address.

**Comment by ymqytw** [comment]
@ahakanbaba I think we should first add the unit test in master then maybe backport to 1.4 and 1.5

File: pkg/kubectl/cmd/apply_test.go (hunk 1 above)
Comment (ahakanbaba) [inline pkg/kubectl/cmd/apply_test.go]: Great suggestion. WIll add that to the test.

File: pkg/kubectl/cmd/apply_test.go (hunk 2 above)
Comment (ahakanbaba) [inline pkg/kubectl/cmd/apply_test.go]: Are you sure ? 
1) I am actually applying the file with Kind: ThirdPartyResource. Not an entry to test-tpr. 
2) This test passes as is.

**Comment by ahakanbaba** [comment]
@ymqytw 
I agree about first merging this to master. 
I think you will need some of the code in this PR to implement other unit tests suggested in https://github.com/kubernetes/kubernetes/pull/40666#issuecomment-276563996 

I plan to work on rebasing this onto master soon. Possibly later today.

**Comment by ymqytw** [comment]
Oh yes. The unit test won't pass until #40666 get in.

**Comment by ahakanbaba** [comment]
Closing this PR in favor of https://github.com/kubernetes/kubernetes/pull/40956. 
There I have added this test onto master. 
There were some changes in the fake fwk setup.

File: pkg/kubectl/cmd/apply_test.go (hunk 2 above)
Comment (ymqytw) [inline pkg/kubectl/cmd/apply_test.go]: You are right. You are applying a TPR definition.

**Pull request closed without merging**